lib/actors/actors.nit

```diff
@@ -0,0 +1,238 @@
+	# i.e. the real working object
+	var instance: V
+
+	# Mailbox used to receive and process messages
+	var mailbox = new BlockingQueue[Message].with_actor(self)
+
+	# Is `self` working ?
+	# i.e. does it have messages to process or is it processing one now ?
+	var working = false
+
+	redef fun main do
+		loop
+			var m = mailbox.shift
+			if m isa ShutDownMessage then
+				sys.active_actors.remove(self)
+				return null
+			end
+			m.invoke(instance)
+			if mailbox.is_empty then
+				working = false
+				sys.active_actors.remove(self)
+			end
+		end
+	end
+
+	# Ends `self`, not responsible if work was ongoing
+	fun kill do
+		var n = self.native
+		if n != null then n.cancel
+	end
+end
+
+# A Message received by a Mailbox
+# In fact, this is the reification of a call
+# Each Message class represent a call to make on `instance` via `invoke`
+abstract class Message
+
+	# Type of the class on which `self` make the call
+	type E: Object
+
+	# Redef this method so it calls the right one on `instance` (double dispatch)
+	fun invoke(instance: E) is abstract
+end
+
+# Abstraction of proxies for threaded actors
+class Proxy
+
+	# Type of the actor `self` is proxiing
+	type E: Actor
+
+	# The proxied actor
+	var actor: E is noinit
+
+	# Kill `actor` without mercy
+	fun kill do actor.kill
+
+	# Tell `actor` to terminate properly
+	# Queueing a ShutDownMessage to the end of its mailbox
+	fun terminate do
+		var msg = new ShutDownMessage
+		actor.mailbox.push(msg)
+	end
+
+	# Tell `actor` to terminate now
+	# Queueing a ShutDownMessage before every other ones
+	fun terminate_now do
+		var msg = new ShutDownMessage
+		actor.mailbox.unshift(msg)
+	end
+
+	# Wait for `actor` to terminate
+	fun wait_termination do actor.join
+end
+
+# A Message to Rule them all... properly shutdown an Actor
+# It's behaviour is implemented in the actor itself
+class ShutDownMessage
+	super Message
+end
+
+# The promise of a value which will be set asynchronously
+class Future[E]
+	# Value promised by `self`
+	var value: nullable E = null
+
+	# Mutex for synchronisation
+	protected var mutex = new Mutex
+
+	# Condition variable for synchronisation
+	protected var cond: nullable PthreadCond = null
+
+	# Can be used to check if the value is available without waiting
+	protected var is_done = false
+
+	# Set the value and signal so that, someone waiting for `value` can retrieve it
+	fun set_value(value: E) do
+		mutex.lock
+		is_done = true
+		self.value = value
+		var cond = self.cond
+		if cond != null then cond.signal
+		mutex.unlock
+	end
+
+	# Return immediatly if `value` is set, or block waiting for `value` to be set
+	fun join: E do
+		mutex.lock
+		if not is_done then
+			var cond = self.cond
+			if cond == null then
+				cond = new PthreadCond
+				self.cond = cond
+			end
+			cond.wait(mutex)
+		end
+		mutex.unlock
+		return value
+	end
+end
+
+# A Blocking queue implemented from a `ConcurrentList`
+# Corresponds to the mailbox of an actor
```

Comment:
I don't really want people to use pop on a BlockingQueue, and this one is specific to my actors.

Ultimately, a future PR would be needed to abstract the concept of a BlockingQueue in the `pthreads` module I think. But there is enough things in this one to report this to a future one.